packages/uni_app/lib/controller/fetchers/course_units_fetcher/course_units_info_fetcher.dart

```diff
@@ -14,31 +14,39 @@ class CourseUnitsInfoFetcher implements SessionDependantFetcher {
     return NetworkRouter.getBaseUrlsFromSession(session).toList();
   }
 
-  Future<CourseUnitSheet> fetchCourseUnitSheet(
-    Session session,
-    int occurrId,
-  ) async {
-    // if course unit is not from the main faculty, Sigarra redirects
-    final url = '${getEndpoints(session)[0]}ucurr_geral.ficha_uc_view';
-    final response = await NetworkRouter.getWithCookies(
-      url,
-      {'pv_ocorrencia_id': occurrId.toString()},
-      session,
-    );
-    return parseCourseUnitSheet(response);
-  }
-
   Future<Sheet> fetchSheet(
     Session session,
     int occurId,
   ) async {
-    final url = '${getEndpoints(session)[0]}mob_ucurr_geral.perfil';
-    final response = await NetworkRouter.getWithCookies(
-      url,
-      {'pv_ocorrencia_id': occurId.toString()},
-      session,
-    );
-    return parseSheet(response);
+    Response? bestResponse;
+
+    for (final endpoint in getEndpoints(session)) {
+      final url = '$endpoint' 'mob_ucurr_geral.perfil';
+      try {
+        final response = await NetworkRouter.getWithCookies(
+          url,
+          {'pv_ocorrencia_id': occurId.toString()},
+          session,
+        );
+
+        if (bestResponse == null ||
+            response.body.length > bestResponse.body.length) {
+          bestResponse = response;
+        }
+      } catch (_) {
+        continue;
+      }
+    }
```

Comment:
```suggestion
    final responses = await Future.wait(
      getEndpoints(session)
        .map((endpoint) => '$endpoint' 'mob_ucurr_geral.perfil')
        .map((url) => NetworkRouter.getWithCookies(
            url,
            {'pv_ocorrencia_id': occurId.toString()},
            session,
          ).catchError((_) => null)
        )
      );

    final bestResponse = responses
      .where((response) => response != null)
      .cast<Response>()
      .fold(null, (best, current) =>
          best == null || current.body.length > best.body.length
            ? current
            : best
      );
```

Reply:
This code needs to be parallelized, otherwise, dart will wait for a request for a curricular unit to complete before proceeding with the other curricular units.
This can be done by doing something like the above. I've also refactored the code to use functional programming so it is more concise and readable.